File: app/src/main/java/com/company/art_and_culture/myarts/museum_fragment/MuseumRepository.java
package com.company.art_and_culture.myarts.museum_fragment;

import android.app.Application;

import androidx.lifecycle.LiveData;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;

import com.company.art_and_culture.myarts.Constants;
import com.company.art_and_culture.myarts.pojo.Art;
import com.company.art_and_culture.myarts.pojo.ArtProvider;
import com.company.art_and_culture.myarts.pojo.Maker;

import java.util.ArrayList;

public class MuseumRepository {

    private LiveData<PagedList<Art>> artList;
    private ArtDataSource artDataSource;
    private MuseumInfoDataSource museumInfoDataSource;
    private ArtDataSourceFactory artDataSourceFactory;
    private Application application;

    private String artProviderId;

    public MuseumRepository(Application application, String artProviderId) {

        this.application = application;
        this.artProviderId = artProviderId;

        PagedList.Config config = new PagedList.Config.Builder()
                .setEnablePlaceholders(false)
                .setPageSize(Constants.PAGE_SIZE)
                .setInitialLoadSizeHint(Constants.PAGE_SIZE)
                .build();
        artDataSourceFactory = new ArtDataSourceFactory(application, artProviderId);
        artList = new LivePagedListBuilder<>(artDataSourceFactory, config).build();

        artDataSource = (ArtDataSource) artDataSourceFactory.create();//if remove this line artLike will not working after refresh

        museumInfoDataSource = new MuseumInfoDataSource(application, artProviderId);
    }

    public LiveData<PagedList<Art>> getArtList(){
        return artList;
    }

    public LiveData<Boolean> getIsLoading() {
        return museumInfoDataSource.getIsLoading();
    }

    public LiveData<Boolean> getIsListEmpty() {
        return artDataSource.getIsListEmpty();
    }

    public LiveData<ArtProvider> getArtProvider() {
        return museumInfoDataSource.getArtProvider();
    }

    public LiveData<Boolean> getArtProviderLike() {
        return museumInfoDataSource.getArtProviderLike();
    }

    public LiveData<ArrayList<Maker>> getListMakers() {
        return museumInfoDataSource.getListMakers();
    }

    public void writeDimentionsOnServer(Art art) {
        artDataSource.writeDimentionsOnServer(art);
    }

    public boolean refresh() {
        museumInfoDataSource.refresh();
        return artDataSourceFactory.refresh();
    }

    public void setArtProviderId(String artProviderId) {

        if (!this.artProviderId.equals(artProviderId)) {
            this.artProviderId = artProviderId;
            artDataSourceFactory.setArtProviderId(artProviderId);
            museumInfoDataSource.setArtProviderId(artProviderId);
        }
    }

    public boolean likeMuseum(ArtProvider museum, String userUniqueId) {
        boolean isConnected = artDataSource.isNetworkAvailable();
        if (isConnected){
            museumInfoDataSource.likeMuseum(museum, userUniqueId);
        }
        return isConnected;
    }
}

File: app/src/main/java/com/company/art_and_culture/myarts/museum_fragment/MuseumRepository.kt
package com.company.art_and_culture.myarts.museum_fragment

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.paging.*
import com.company.art_and_culture.myarts.Constants
import com.company.art_and_culture.myarts.network.MainHttpService
import com.company.art_and_culture.myarts.network.MuseumArtsService
import com.company.art_and_culture.myarts.pojo.Art
import com.company.art_and_culture.myarts.pojo.ArtProvider
import com.company.art_and_culture.myarts.pojo.Maker
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.withContext

class MuseumRepository(
    private val artProviderId: String,
    private val userUniqueId: String,
    private val ioDispatcher: CoroutineDispatcher
) : IMuseumRepository {

    private var museumInfoDataSource: MuseumInfoDataSource = MuseumInfoDataSource(MainHttpService.create(), artProviderId, userUniqueId)
    private var museumArtsDataSource: MuseumArtsDataSource = MuseumArtsDataSource(MuseumArtsService.create(), artProviderId, userUniqueId)

    override suspend fun getArtsFlow(artProviderId: String): Flow<PagingData<Art>>
        = withContext(ioDispatcher) {

            museumArtsDataSource = MuseumArtsDataSource(MuseumArtsService.create(), artProviderId, userUniqueId)
            museumInfoDataSource.setMuseumId(artProviderId)

            return@withContext Pager(
                config = PagingConfig(
                    pageSize = Constants.PAGE_SIZE,
                    enablePlaceholders = true
                ),
                pagingSourceFactory = { museumArtsDataSource }
            ).flow
        }

    override fun onDestroy() {
        museumInfoDataSource.onDestroy()
    }

    override fun getIsLoading(): LiveData<Boolean> {
        return museumInfoDataSource.isLoading
    }

    override fun isContentEmpty(): LiveData<Boolean> {
        return museumInfoDataSource.isContentEmpty
    }

    override fun getArtProvider(): MutableLiveData<ArtProvider?> {
        return museumInfoDataSource.artProvider
    }

    override fun getIsArtProviderLiked(): MutableLiveData<Boolean?> {
        return museumInfoDataSource.isArtProviderLiked
    }

    override fun getListMakers(): MutableLiveData<ArrayList<Maker>?> {
        return museumInfoDataSource.listMakers
    }

    override suspend fun writeDimensionsOnServer(art: Art?) =
        withContext(ioDispatcher) {
            museumInfoDataSource.writeDimensionsOnServer(art)
        }

    override suspend fun likeMuseum(museum: ArtProvider?, userUniqueId: String?) =
        withContext(ioDispatcher) {
            museumInfoDataSource.likeMuseum(museum, userUniqueId)
        }


}